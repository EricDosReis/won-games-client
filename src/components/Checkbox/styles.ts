import styled, { css } from 'styled-components';

import { CheckboxProps } from '.';

export const Wrapper = styled.div`
  display: flex;
  align-items: center;
`;

export const Input = styled.input`
  ${({ theme }) => css`
    appearance: none;
    border: 0.2rem solid ${theme.colors.darkGray};
    border-radius: ${theme.border.radius};
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
    height: 1.8rem;
    width: 1.8rem;
    outline: none;
    position: relative;
    transition: background-color border ${theme.transition.fast};

    &::before {
      content: '';
      border: 0.2rem solid ${theme.colors.white};
      border-top: 0;
      border-left: 0;
      height: 0.9rem;
      width: 0.6rem;
      opacity: 0;
      position: absolute;
      top: 0.1rem;
      transform: rotate(45deg);
      transition: ${theme.transition.fast};
    }

    &:focus {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    &:checked {
      background: ${theme.colors.primary};
      border-color: ${theme.colors.primary};

      &:before {
        opacity: 1;
      }
    }
  `}
`;

export const Label = styled.label<Pick<CheckboxProps, 'labelColor'>>`
  ${({ theme, labelColor }) => css`
    color: ${theme.colors[labelColor!]};
    cursor: pointer;
    line-height: 1.7rem;
    padding-left: ${theme.spacings.xxsmall};
  `}
`;
