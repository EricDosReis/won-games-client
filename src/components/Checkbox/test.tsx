import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithTheme } from 'utils/tests/helpers';

import Checkbox from '.';

describe('<Checkbox />', () => {
  it('should render the checkbox', () => {
    const { container } = renderWithTheme(
      <Checkbox label="Checkbox" labelFor="check" />,
    );

    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(screen.getByLabelText(/checkbox/i)).toBeInTheDocument();
    expect(screen.getByText(/checkbox/i)).toHaveAttribute('for', 'check');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render without label', () => {
    renderWithTheme(<Checkbox />);

    expect(screen.queryByLabelText(/checkbox/i)).not.toBeInTheDocument();
  });

  it('should render with black label', () => {
    renderWithTheme(
      <Checkbox label="Checkbox" labelFor="check" labelColor="black" />,
    );

    expect(screen.getByText(/checkbox/i)).toHaveStyle({
      color: '#030517',
    });
  });

  it('should dispatch onCheck when status changes', async () => {
    const onCheck = jest.fn();

    renderWithTheme(<Checkbox label="Checkbox" onCheck={onCheck} />);

    expect(onCheck).not.toBeCalled();

    userEvent.click(screen.getByRole('checkbox'));

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toHaveBeenCalledWith(true);
  });

  it('should dispatch onCheck when status changes', async () => {
    const onCheck = jest.fn();

    renderWithTheme(<Checkbox label="Checkbox" onCheck={onCheck} isChecked />);

    userEvent.click(screen.getByRole('checkbox'));

    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toHaveBeenCalledWith(false);
  });

  it('should be accessible with tab', () => {
    renderWithTheme(<Checkbox label="Checkbox" labelFor="Checkbox" />);

    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(screen.queryByLabelText(/checkbox/i)).toHaveFocus();
  });
});
