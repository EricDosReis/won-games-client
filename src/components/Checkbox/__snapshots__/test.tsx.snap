// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render the checkbox 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 0.2rem solid #2E2F42;
  border-radius: 0.4rem;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 1.8rem;
  width: 1.8rem;
  outline: none;
  position: relative;
  -webkit-transition: background-color border 0.1s ease-in-out;
  transition: background-color border 0.1s ease-in-out;
}

.c1::before {
  content: '';
  border: 0.2rem solid #FAFAFA;
  border-top: 0;
  border-left: 0;
  height: 0.9rem;
  width: 0.6rem;
  opacity: 0;
  position: absolute;
  top: 0.1rem;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
}

.c1:focus {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c1:checked {
  background: #F231A5;
  border-color: #F231A5;
}

.c1:checked:before {
  opacity: 1;
}

.c2 {
  color: #FAFAFA;
  cursor: pointer;
  line-height: 1.7rem;
  padding-left: 0.8rem;
}

<div
  class="c0"
>
  <input
    class="c1"
    id="check"
    type="checkbox"
    value=""
  />
  <label
    class="c2"
    for="check"
  >
    Checkbox
  </label>
</div>
`;
